strawberryfields/engine.py

```diff
@@ -86,16 +87,45 @@ class Result:
 
     Represents the results of the execution of a quantum program.
     Returned by :meth:`.BaseEngine.run`.
+
+    Args:
+        samples (dict[int, List[int, float]): dictionary of measurement samples;
+            keys are mode numbers, entries have shape ``(modes, shots)``
     """
     def __init__(self, samples):
         #: BaseState: quantum state object returned by a local backend, if any
         self.state = None
-        #: array(array(Number)): measurement samples, shape == (modes,) or shape == (shots, modes)
-        # ``samples`` arrives as a list of arrays, need to convert here to a multidimensional array
-        if len(shape(samples)) > 1:
-            samples = stack(samples, 1)
+
+        #: dict[int, List[int, float]]
         self.samples = samples
 
+        #: list[int]: list of measured modes
+        self.measured_modes = list(samples.keys())
+
+        # samples arrives as a dictionary, need to convert here to a multidimensional array
+        if len(samples) == 0:
+            samples_array = np.array([])
+        else:
+            arrs = [np.array(v) for v in samples.values()]
+            samples_shapes = [v.shape for v in arrs]
+            ref_shape = samples_shapes[0]
+
+            if not np.all([s == ref_shape for s in samples_shapes]):
+                raise ValueError("The returned samples do not all have the same shape.")
+
+            if len(ref_shape) == 0:
+                #  scalars, single shot
+                samples_array = np.expand_dims(np.stack(arrs), axis=1)  # shape==(modes, 1)
```

Comment:
Isn't the sample data already always in a list, even if shots==1? This case should not be necessary then.